crates/accelerate/src/sparse_observable.rs

```diff
@@ -0,0 +1,1677 @@
+    #[pyo3(signature = (data, /, num_qubits=None))]
+    #[new]
+    fn py_new(data: Bound<PyAny>, num_qubits: Option<u32>) -> PyResult<Self> {
+        let py = data.py();
+        let check_num_qubits = |data: &Bound<PyAny>| -> PyResult<()> {
+            let Some(num_qubits) = num_qubits else {
+                return Ok(());
+            };
+            let other_qubits = data.getattr(intern!(py, "num_qubits"))?.extract::<u32>()?;
+            if num_qubits == other_qubits {
+                return Ok(());
+            }
+            Err(PyValueError::new_err(format!(
+                "explicitly given 'num_qubits' ({num_qubits}) does not match operator ({other_qubits})"
+            )))
+        };
+
+        if data.is_instance(PAULI_TYPE.get_bound(py))? {
+            check_num_qubits(&data)?;
+            return Self::py_from_pauli(data);
+        }
+        if data.is_instance(SPARSE_PAULI_OP_TYPE.get_bound(py))? {
+            check_num_qubits(&data)?;
+            return Self::py_from_sparse_pauli_op(data);
+        }
+        if let Ok(label) = data.extract::<String>() {
+            let num_qubits = num_qubits.unwrap_or(label.len() as u32);
+            if num_qubits as usize != label.len() {
+                return Err(PyValueError::new_err(format!(
+                    "explicitly given 'num_qubits' ({}) does not match label ({})",
+                    num_qubits,
+                    label.len(),
+                )));
+            }
+            return Self::py_from_label(&label).map_err(PyErr::from);
+        }
+        if let Ok(observable) = data.downcast::<Self>() {
+            check_num_qubits(&data)?;
+            return Ok(observable.borrow().clone());
+        }
+        if let Ok(vec) = data.extract() {
+            return Self::py_from_list(vec, num_qubits);
+        }
+        if let Ok(vec) = data.extract() {
+            let Some(num_qubits) = num_qubits else {
+                return Err(PyValueError::new_err(
+                    "if using the sparse-list form, 'num_qubits' must be provided",
+                ));
+            };
+            return Self::py_from_sparse_list(vec, num_qubits).map_err(PyErr::from);
+        }
+        Err(PyTypeError::new_err(format!(
+            "unknown input format for 'SparseObservable': {}",
+            data.get_type().repr()?,
+        )))
+    }
+
+    /// The number of qubits the operator acts on.
+    ///
+    /// This is not inferable from any other shape or values, since identities are not stored
+    /// explicitly.
+    #[getter]
+    #[inline]
+    pub fn num_qubits(&self) -> u32 {
+        self.num_qubits
+    }
+
+    /// The number of terms in the sum this operator is tracking.
+    #[getter]
+    #[inline]
+    pub fn num_ops(&self) -> usize {
+        self.coeffs.len()
+    }
+
+    /// The coefficients of each abstract term in in the sum.  This has as many elements as terms in
+    /// the sum.
+    #[getter]
+    fn get_coeffs(slf_: Py<Self>) -> ArrayView {
```

Comment:
You can think of the PyO3 `#[getter]` tag as being give-or-take the same thing as the Python-space `@property` decorator - it causes the function following it to be used for base attribute access, and doesn't define a setter.  In addition, though, PyO3 also automatically strips any `get_` prefix from the front of a getter - that's useful in Rust if you have both a getter _and_ a setter, because in Rust you can't name two methods the same thing.

I used the `get_` suffix here to avoid a naming conflict in Rust with the general `coeffs` field, and with a view that it feels quite likely that at some point in the future we'll make a `pub fn coeffs(&self) -> &[Complex64]` view method.

Reply:
Oh right, that makes sense -- thanks for the info!